File: app/src/main/java/com/myapplication/rick/packagedelivery/Street.java
package com.myapplication.rick.packagedelivery;

import android.support.annotation.NonNull;

import java.util.Locale;
/**
 * Created by Rick on 1-5-2016.
 */
class Street implements CSVWriteAble {
    private String name;
    private Range range;
    private Direction direction;

    public Street(@NonNull String name, @NonNull Range range) throws IllegalArgumentException {
        if (name == null || name.trim().length() < 1) {
            throw new IllegalArgumentException("Invalid name");
        }
        if (range == null) {
            throw new IllegalArgumentException("Invalid range");
        }

        this.name = name;
        this.range = range;
        this.direction = Direction.LowToHigh; //default value
    }

    public Street(@NonNull String name, @NonNull Range range, @NonNull Direction direction) throws IllegalArgumentException {
        if (name == null || name.trim().length() < 1) {
            throw new IllegalArgumentException("Invalid name");
        }
        if (range == null) {
            throw new IllegalArgumentException("Invalid range");
        }
        if (direction == null) {
            throw new IllegalArgumentException("Invalid direction");
        }

        this.name = name;
        this.range = range;
        this.direction = direction;
    }

    @Override
    public String toString() {
        return String.format(Locale.US, "%s, %s", name, range.toString());
    }

    public String[] toCSV() {
        String[] rangeValues = range.toCSV();
        String[] values = new String[1 + range.length()];
        values[0] = name;
        //should be same as arraycopy
        //for (int i = 0; i < range.length(); i++) {
        //values[i + 1] = rangeValues[i];
        //}
        System.arraycopy(rangeValues, 0, values, 1, rangeValues.length);
        return values;
    }

    public Range getRange() {
        return range;
    }

    public String getName() {
        return name;
    }

    //public Direction getDirection() {
    //    return direction;
    //}
}

File: app/src/main/java/com/myapplication/rick/packagedelivery/Street.kt
package com.myapplication.rick.packagedelivery

import android.os.Parcel
import android.os.Parcelable
import java.util.Locale

/**
 * Created by Rick on 1-5-2016.
 */
class Street(val name: String, val range: Range, val direction: Direction = Direction.LowToHigh) : CSVWriteAble, Parcelable {
    init {
        if (name.isBlank()) throw IllegalArgumentException("Empty name")
    }

    override fun toString(): String = "$name, $range"

    fun isValidNumber(number: Int): Boolean {
        val evenNumber = number % 2 == 0
        val correspondsWithRangeType = (range.rangeType == RangeType.All) ||
                (evenNumber && (range.rangeType == RangeType.Even)) ||
                (!evenNumber && (range.rangeType == RangeType.Uneven))
        val inRange = number >= range.lowerBound &&
                number <= range.upperBound
        return correspondsWithRangeType && inRange
    }

    override fun toCSV(): Array<String> {
        val values = ArrayList(range.toCSV().asList())
        values.add(0, name)
        values.add(direction.toString())
        return values.toTypedArray()
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(name)
        parcel.writeParcelable(range, flags)
        parcel.writeParcelable(direction, flags)
    }

    override fun describeContents(): Int = 0

    companion object {
        @JvmField
        val CREATOR = object : Parcelable.Creator<Street> {
            override fun newArray(size: Int): Array<Street?> = arrayOfNulls<Street?>(size)

            override fun createFromParcel(parcel: Parcel): Street {
                val name: String = parcel.readString()
                val range: Range = parcel.readParcelable(Range::class.java.classLoader)
                val direction: Direction = parcel.readParcelable(Direction::class.java.classLoader)
                return Street(name, range, direction)
            }
        }
    }
}
